subscriptions/fixed-price/checkout/node/src/server.ts

```diff
@@ -207,3 +247,248 @@ app.post('/webhook', async (req, res) => {
+      if (subscriptionInDatabase.status === 'trailing') {
+        
+        // Step 1: Update subscription to active 
+        SubscriptionTable.update(subscriptionInDatabase.id, { status: 'active' })
+
+        // Step 2: Optionally send invoice and store charges locally
+        // ...
+
+        return
+      }
+        
+    ```
+  */
+
+  // Handle past due charge successful (S6 ✅)
+  if (charge.attempt_count > 1) {
+    /*    
+      ```
+        // Step 1: Update subscription to active 
+        SubscriptionTable.update(subscriptionInDatabase.id, { status: 'active' })
+
+        // Step 2: Optionally send invoice and store charges locally
+        // ...
+
+        return
+      ```
+    */
+  }
+  switch (charge.reason) {
+    case 'subscription_creation': // S1 ✅
+    case 'subscription_due': // S2 ✅
+    case 'subscription_update': // S4 ✅
+    case 'subscription_cancel': // S5 ✅
+      /*
+        You may want to do the following:
+        - Send an invoice to your customer.
+        - Store charges in your DB.
+      */
+      break
+
+    default:
+      break
+  }
+}
+
+/**
+ * Subscription created
+ *
+ * We recommend pivoting on `subscription.created`, to handle the case when a checkout is a completed.
+ * Use this event to provide feedback though the UI, and create a new subscription in your database.
+ *
+ * @param subscription The subscription object received in the event
+ */
+async function handleSubscriptionCreated(
+  subscription: DiagonalSubscription,
+): Promise<void> {
+  /*
+      Find the user in your database using the `customer_id` or `reference`, specified on the subscription object:
+
+      ```
+        const user = UserTable.findOne({
+          diagonalCustomerId: subscription.customer_id!,
+        })
+      ```
+
+      Create a new subscription in your database:
+    
+      ```
+        SubscriptionTable.create({
+          userId: user.id,
+          status: 'active',
+          diagonalSubscriptionId: subscription.id,
+          planId: subscription.reference,
+        })
+      ```
+
+  */
+}
+
+/**
+ * Charge attempt failed
+ *
+ * @param charge The charge object received in the event
+ */
+async function handleChargeAttemptFailed(charge: Charge): Promise<void> {
+  /*
+    You may want to notify your user that the subscription charge failed, 
+    based on the failure reason specified in `charge.last_attempt_failure_reason`
+
+    NOTE: charge will be re-attempted at `charge.next_attempt_at`
+  */
+
+  switch (charge.last_attempt_failure_reason) {
+    case 'insufficient_allowance':
+      // Notify the user to increase their spending allowance on subscriptions.diagonal.finance
+      break
+    case 'insufficient_balance':
+      // Notify the user to fund their wallet
+      break
+    default:
+      break
+  }
+}
+
+/**
+ * Charge failed
+ *
+ *  We recommend pivoting on `charge.failed` when handling the following subscription lifecycle events:
+ *
+ *   - S1: Subscription creation failed
```

Comment:
Charge can fail when the address is blacklisted, can we include that as well?

Reply:
Since subscription is automatically cancelled when a blacklisted address is used, we recommend handle that logic in the `subscription.canceled` event. 

Did you see the line just below this one?

> `charge.failed` can fire in other scenarios e.g. "maximum number of retry attempts reached" or "address blacklist usdc". We recommend handling these scenarios in `handleSubscriptionCanceled`.